proving/proving.go

```diff
@@ -56,26 +60,68 @@ func (p *Prover) generateProof(id []byte, challenge Challenge) (*Proof, error) {
 	proof.Challenge = challenge
 	proof.Identity = id
 
-	dir := shared.GetInitDir(p.cfg.DataDir, id)
-	reader, err := persistence.NewLabelsReader(dir)
+	result, err := p.generateMTree(id, challenge)
+	if err != nil {
+		return nil, err
+	}
+	proof.MerkleRoot = result.Root
+
+	leafReader := result.Reader.GetLayerReader(0)
+	width, err := leafReader.Width()
+	if err != nil {
+		return nil, err
+	}
+
+	provenLeafIndices := CalcProvenLeafIndices(
+		proof.MerkleRoot, width<<difficulty, uint8(p.cfg.NumOfProvenLabels), difficulty)
+
+	_, proof.ProvenLeaves, proof.ProofNodes, err = merkle.GenerateProof(provenLeafIndices, result.Reader)
+	if err != nil {
+		return nil, err
+	}
+
+	err = leafReader.Close()
 	if err != nil {
 		return nil, err
 	}
-	width, err := reader.Width()
+
+	return proof, nil
+}
+
+func (p *Prover) generateMTree(id []byte, challenge Challenge) (*MTreeOutput, error) {
+	dir := shared.GetInitDir(p.cfg.DataDir, id)
+	readers, err := persistence.GetReaders(dir)
 	if err != nil {
 		return nil, err
 	}
-	if width*difficulty.LabelsPerGroup() >= math.MaxUint64 {
-		return nil, fmt.Errorf("leaf reader too big, number of label groups (%d) * labels per group (%d) "+
-			"overflows uint64", width, difficulty.LabelsPerGroup())
+
+	parallelism := p.CalcParallelism(len(readers))
+	if parallelism > 1 {
```

Comment:
I wonder if the existing method for "no parallelism" is more efficient than running the new method with a single worker.

I have a suspicion that the new method is actually more performant, but even if it's about the same -- removing this `if` and the old method would simplify the code and reduce the possibility for hidden bugs (better to have a single code path for a certain task).

I suggest benchmarking this to decide.

Reply:
Good point.. 
The new method with a single worker is indeed a bit more performant.